File: src/main/java/com/HOIIVUtils/clauzewitz/script/CollectionPDXScript.java
package com.HOIIVUtils.clauzewitz.script;

import com.HOIIVUtils.clausewitz_parser.Node;
import com.HOIIVUtils.clausewitz_parser.NodeValue;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.function.Consumer;
import java.util.function.Supplier;
import java.util.stream.Stream;

abstract public class CollectionPDXScript<T extends PDXScript<?>> extends AbstractPDX<List<T>> implements Iterable<T> {

    public CollectionPDXScript(@NotNull String... pdxIdentifiers) {
        super(pdxIdentifiers);
        obj = new ArrayList<>();
    }

    public CollectionPDXScript(@NotNull List<String> pdxIdentifiers) {
        super(pdxIdentifiers);
        obj = new ArrayList<>();
    }

    @Override
    public void loadPDX(Node expression) throws UnexpectedIdentifierException {
        try {
            add(expression);
        } catch (NodeValueTypeException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void loadPDX(List<Node> expressions) {
        if (expressions == null) return;
        expressions.stream().filter(this::isValidIdentifier)
                .forEach(expression -> {
                    try {
                        loadPDX(expression);
                    } catch (UnexpectedIdentifierException e) {
                        System.err.println(e.getMessage());
                        //throw new RuntimeException(e);
                    }
                });
    }

    @Override
    public boolean objEquals(PDXScript<?> other) {
        return false; // todo? well.
    }

    @Override
    public List<T> get() {
        return super.get();
    }


    protected void add(Node expression) throws UnexpectedIdentifierException, NodeValueTypeException {
        //usingIdentifier(expression);  // could be any identifier based on T
        NodeValue value = expression.value();

        // if this PDXScript is an encapsulation of PDXScripts (such as Focus)
        // then load each sub-PDXScript
        if (obj instanceof ArrayList<T> childScriptList) {
            T childScript = newChildScript(expression);
            childScript.loadPDX(expression);
            childScriptList.add(childScript);
        } else {
            try {
                obj.add((T) value.valueObject());
            } catch (ClassCastException e) {
                throw new NodeValueTypeException(expression, e);
            }
        }
    }

    abstract protected T newChildScript(Node expression);

    public void clear() {
        obj.clear();
    }

    public boolean isEmpty() {
        return get().isEmpty();
    }

    @NotNull
    @Override
    public Iterator<T> iterator() {
        return get().iterator();
    }

    @Override
    public void forEach(Consumer<? super T> action) {
        get().forEach(action);
    }

    @Override
    public Spliterator<T> spliterator() {
        return get().spliterator();
    }

    public int size() {
        return get().size();
    }

    public Stream<T> stream() {
        return get().stream();
    }

    @Override
    public boolean isUndefined() {
        return obj.isEmpty();
    }

    @Override
    public @Nullable String toScript() {
        StringBuilder sb = new StringBuilder();
        var scripts = get();
        if (scripts == null) return null;
        for (T pdxScript : scripts) {
            var str = pdxScript.toScript();
            if (str == null) continue;
            sb.append(str);
        }
        return sb.toString();
    }
}

File: src/main/scala/com/hoi4utils/clausewitz/script/CollectionPDXScript.scala
package com.hoi4utils.clausewitz.script

import com.hoi4utils.clausewitz_parser.Node
import com.hoi4utils.clausewitz_parser.NodeValue
import org.jetbrains.annotations.NotNull
import org.jetbrains.annotations.Nullable

import java.util.function.Consumer
import java.util.stream.Stream

// todo i do not like this class
abstract class CollectionPDXScript[T <: PDXScript[?]](pdxIdentifiers: String*) extends AbstractPDX[ListBuffer[T]](pdxIdentifiers) with Iterable[T] {
  def this(pdxIdentifiers: List[String]) = {
    this(pdxIdentifiers*)
  }

  @throws[UnexpectedIdentifierException]
  override def loadPDX(expression: Node): Unit = {
    try add(expression)
    catch {
      case e: NodeValueTypeException =>
        throw new RuntimeException(e)
    }
  }

  override def loadPDX(expressions: List[Node]): Unit = {
    if (expressions != null)
      expressions.stream.filter(this.isValidIdentifier).forEach((expression: Node) => {
        try loadPDX(expression)
        catch {
          case e: UnexpectedIdentifierException =>
            System.err.println(e.getMessage)
          //throw new RuntimeException(e);
        }

      })
  }

  override def nodeEquals(other: PDXScript[?]) = false // todo? well.

  override def get(): ListBuffer[T] = super.get()

  @throws[UnexpectedIdentifierException]
  @throws[NodeValueTypeException]
  protected def add(expression: Node): Unit = {
    //usingIdentifier(expression);  // could be any identifier based on T
    val value = expression.value
    // if this PDXScript is an encapsulation of PDXScripts (such as Focus)
    // then load each sub-PDXScript
    node.$ match {
      case _: ListBuffer[T] =>
        val childScript = newChildScript(expression)
        childScript.loadPDX(expression)
        childScriptList.add(childScript)
      case _ =>
        // todo?
    }
  }

  protected def newChildScript(expression: Node): T

  def clear(): Unit = {
    node.$.clear
  }

  def isEmpty: Boolean = get().isEmpty

  override def iterator: Iterator[T] = get().iterator

  override def forEach(action: Consumer[? >: T]): Unit = {
    get().foreach(action)
  }

//  override def spliterator: java.util.Spliterator[T] = get().spliterator

  def size: Int = get().size

//  def stream: Stream[T] = get().stream

  override def isUndefined: Boolean = node.isEmpty

  override def toScript: String = {
    val sb = new StringBuilder
    val scripts = get()
    if (scripts == null) return null
    for (pdxScript <- scripts) {
      val str = pdxScript.toScript
      if (str == null) continue //todo: continue is not supported
      sb.append(str)
    }
    sb.toString
  }
}


